Panic on slice bounds out of range

**Describe the bug**
Telepresence intercept terminates unexpectedly, and this panic is seen in the daemon.log:

panic: runtime error: slice bounds out of range [:28275] with capacity 1520
goroutine 845 [running]:
github.com/telepresenceio/telepresence/v2/pkg/vif/udp.Header.Payload(...)
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/header.go:57
github.com/telepresenceio/telepresence/v2/pkg/vif/udp.(*handler).writeLoop(0xc000025b60, {0x2d72228, 0xc0004ebc80})
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/handler.go:122 +0x6be
created by github.com/telepresenceio/telepresence/v2/pkg/vif/udp.(*handler).Start
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/handler.go:99 +0xbd

I'm having the same issue, were you able to resolve or work around this issue @chongkai ?

```
panic: runtime error: slice bounds out of range [:26990] with capacity 1520

goroutine 202 [running]:
github.com/telepresenceio/telepresence/v2/pkg/vif/udp.Header.Payload(...)
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/header.go:57
github.com/telepresenceio/telepresence/v2/pkg/vif/udp.(*handler).writeLoop(0x140002816e0, {0x10405a480, 0x14000644900})
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/handler.go:122 +0x698
created by github.com/telepresenceio/telepresence/v2/pkg/vif/udp.(*handler).Start
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/handler.go:99 +0x88
```

I'm having the same issue, were you able to resolve or work around this issue @chongkai ?

```
panic: runtime error: slice bounds out of range [:26990] with capacity 1520

goroutine 202 [running]:
github.com/telepresenceio/telepresence/v2/pkg/vif/udp.Header.Payload(...)
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/header.go:57
github.com/telepresenceio/telepresence/v2/pkg/vif/udp.(*handler).writeLoop(0x140002816e0, {0x10405a480, 0x14000644900})
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/handler.go:122 +0x698
created by github.com/telepresenceio/telepresence/v2/pkg/vif/udp.(*handler).Start
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/handler.go:99 +0x88
```

I have this problem too.
```
panic: runtime error: slice bounds out of range [:28015] with capacity 1520

goroutine 271 [running]:
github.com/telepresenceio/telepresence/v2/pkg/vif/udp.Header.Payload(...)
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/header.go:57
github.com/telepresenceio/telepresence/v2/pkg/vif/udp.(*handler).writeLoop(0xc0007605f0, {0x5e76720, 0xc000399580})
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/handler.go:106 +0x531
created by github.com/telepresenceio/telepresence/v2/pkg/vif/udp.(*handler).Start
        github.com/telepresenceio/telepresence/v2/pkg/vif/udp/handler.go:87 +0xbd
```

I get it randomly. When try to get some data from external service such us db or http service.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 # Changelog
 
+### 2.6.5 (TBD)
+
+- Bugfix: UDP based communication with services in the cluster now works as expected.
+
 ### 2.6.4 (May 23, 2022)
 
 - Bugfix: The traffic-manager RBAC grants permissions to update services, deployments, replicatsets, and statefulsets. Those

--- a/integration_test/itest/cluster.go
+++ b/integration_test/itest/cluster.go
@@ -129,6 +129,7 @@ func (s *cluster) tearDown(ctx context.Context) {
 	s.ensureQuitAndLoggedOut(ctx)
 	if s.kubeConfig != "" {
 		_ = Run(ctx, "kubectl", "delete", "-f", filepath.Join(s.moduleRoot, "k8s", "client_rbac.yaml"))
+		_ = Run(ctx, "kubectl", "delete", "--wait=false", "ns", "-l", "purpose=tp-cli-testing")
 	}
 }
 

--- a/integration_test/testdata/udp-echo/Dockerfile
+++ b/integration_test/testdata/udp-echo/Dockerfile
@@ -1,0 +1,10 @@
+FROM golang:alpine3.15 AS builder
+
+WORKDIR /udp-echo
+COPY go.mod .
+COPY main.go .
+RUN go build -o udp-echo .
+
+FROM alpine:3.15
+COPY --from=builder /udp-echo/udp-echo /usr/local/bin/
+ENTRYPOINT ["/usr/local/bin/udp-echo"]

--- a/integration_test/testdata/udp-echo/go.mod
+++ b/integration_test/testdata/udp-echo/go.mod
@@ -1,0 +1,3 @@
+module local
+
+go 1.17

--- a/integration_test/testdata/udp-echo/main.go
+++ b/integration_test/testdata/udp-echo/main.go
@@ -1,0 +1,66 @@
+package main
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"sync"
+)
+
+func main() {
+	portsEnv := os.Getenv("PORTS")
+	if portsEnv == "" {
+		portsEnv = os.Getenv("PORT")
+	}
+	if portsEnv == "" {
+		portsEnv = "8080"
+	}
+	ports := strings.Split(portsEnv, ",")
+	wg := sync.WaitGroup{}
+	wg.Add(len(ports))
+	for _, port := range ports {
+		port := port // pin it
+		go func() {
+			defer wg.Done()
+			fmt.Printf("UDP-echo server listening on port %s.\n", port)
+			defer fmt.Printf("UDP-echo server on port %s exited.\n", port)
+			pc, err := net.ListenPacket("udp", ":"+port)
+			if err == nil {
+				err = serveConnection(pc)
+			}
+			if err != nil {
+				fmt.Fprintln(os.Stderr, err.Error())
+			}
+		}()
+	}
+	wg.Wait()
+}
+
+func serveConnection(pc net.PacketConn) error {
+	buf := [0x10000]byte{}
+	pfx := []byte("Reply from UDP-echo: ")
+	for {
+		n, addr, err := pc.ReadFrom(buf[:])
+		if n > 0 {
+			sb := string(buf[:n])
+			if strings.HasSuffix(sb, "\n") {
+				fmt.Print(sb)
+			} else {
+				fmt.Println(sb)
+			}
+			if n == 5 && sb == "exit\n" {
+				return nil
+			}
+			r := make([]byte, len(pfx)+n)
+			copy(r, pfx)
+			copy(r[len(pfx):], buf[:n])
+			if _, werr := pc.WriteTo(r, addr); werr != nil {
+				fmt.Fprintln(os.Stderr, werr.Error())
+			}
+		}
+		if err != nil {
+			return err
+		}
+	}
+}

--- a/integration_test/udp_test.go
+++ b/integration_test/udp_test.go
@@ -3,6 +3,7 @@ package integration_test
 import (
 	"fmt"
 	"net"
+	"runtime"
 	"strings"
 	"time"
 )
@@ -35,8 +36,18 @@ func (s *connectedSuite) TestUDPEcho() {
 
 	mb := strings.Builder{}
 	mb.WriteString("This is ")
-	for i := 0; i < 1000; i++ {
-		mb.WriteString("a russian doll containing ")
+	itm := "a russian doll containing "
+	count := 1000
+	switch runtime.GOOS {
+	case "darwin":
+		// Max UDP message size is 9216 bytes
+		count = 9000 / len(itm)
+	case "windows":
+		// SO_MAX_MSG_SIZE == 8192
+		count = 8000 / len(itm)
+	}
+	for i := 0; i < count; i++ {
+		mb.WriteString(itm)
 	}
 	mb.WriteString("a solid russian doll")
 

--- a/integration_test/udp_test.go
+++ b/integration_test/udp_test.go
@@ -1,0 +1,62 @@
+package integration_test
+
+import (
+	"fmt"
+	"net"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/telepresenceio/telepresence/v2/integration_test/itest"
+)
+
+func (s *connectedSuite) TestUDPEcho() {
+	ctx := s.Context()
+	require := s.Require()
+	svc := "udp-echo"
+	tag := s.Registry() + "/udp-test:" + strings.TrimPrefix(s.TelepresenceVersion(), "v")
+	testDir := filepath.Join("testdata", svc)
+
+	require.NoError(itest.Run(ctx, "docker", "build", "-t", tag, testDir))
+	require.NoError(itest.Run(ctx, "docker", "push", tag))
+	require.NoError(s.Kubectl(ctx, "create", "deploy", svc, "--image", tag))
+	require.NoError(s.Kubectl(ctx, "expose", "deploy", svc, "--port", "80", "--protocol", "UDP", "--target-port", "8080"))
+	defer func() {
+		_ = s.Kubectl(ctx, "delete", "svc,deploy", svc)
+	}()
+	require.NoError(s.RolloutStatusWait(ctx, "deploy/"+svc))
+	require.Eventually(
+		func() bool {
+			stdout := itest.TelepresenceOk(ctx, "list", "--namespace", s.AppNamespace())
+			return strings.Contains(stdout, svc)
+		},
+		12*time.Second, // waitFor
+		3*time.Second,  // polling interval
+		`never included in list output`)
+
+	mb := strings.Builder{}
+	mb.WriteString("This is ")
+	for i := 0; i < 1000; i++ {
+		mb.WriteString("a russian doll containing ")
+	}
+	mb.WriteString("a solid russian doll")
+	conn, err := net.Dial("udp", fmt.Sprintf("%s.%s:80", svc, s.AppNamespace()))
+	require.NoError(err)
+
+	buf := [0x10000]byte{}
+
+	echoTest := func(msg string) {
+		_, err = conn.Write([]byte(msg))
+		require.NoError(err)
+		require.NoError(conn.SetReadDeadline(time.Now().Add(5 * time.Second)))
+		n, err := conn.Read(buf[:])
+		require.NoError(err)
+		rp := "Reply from UDP-echo: "
+		pl := len(rp)
+		require.Equal(string(buf[:pl]), rp)
+		require.Equal(len(msg)+pl, n)
+		require.Equal(msg, string(buf[pl:n]))
+	}
+	echoTest("Hello")
+	echoTest(mb.String())
+}

--- a/integration_test/udp_test.go
+++ b/integration_test/udp_test.go
@@ -25,14 +25,19 @@ func (s *connectedSuite) TestUDPEcho() {
 		_ = s.Kubectl(ctx, "delete", "svc,deploy", svc)
 	}()
 	require.NoError(s.RolloutStatusWait(ctx, "deploy/"+svc))
+
+	var conn net.Conn
 	require.Eventually(
 		func() bool {
-			stdout := itest.TelepresenceOk(ctx, "list", "--namespace", s.AppNamespace())
-			return strings.Contains(stdout, svc)
+			var err error
+			conn, err = net.Dial("udp", fmt.Sprintf("%s.%s:80", svc, s.AppNamespace()))
+			return err == nil
 		},
 		12*time.Second, // waitFor
 		3*time.Second,  // polling interval
-		`never included in list output`)
+		`dial never succeeds`)
+
+	defer conn.Close()
 
 	mb := strings.Builder{}
 	mb.WriteString("This is ")
@@ -40,13 +45,11 @@ func (s *connectedSuite) TestUDPEcho() {
 		mb.WriteString("a russian doll containing ")
 	}
 	mb.WriteString("a solid russian doll")
-	conn, err := net.Dial("udp", fmt.Sprintf("%s.%s:80", svc, s.AppNamespace()))
-	require.NoError(err)
 
 	buf := [0x10000]byte{}
 
 	echoTest := func(msg string) {
-		_, err = conn.Write([]byte(msg))
+		_, err := conn.Write([]byte(msg))
 		require.NoError(err)
 		require.NoError(conn.SetReadDeadline(time.Now().Add(5 * time.Second)))
 		n, err := conn.Read(buf[:])

--- a/integration_test/udp_test.go
+++ b/integration_test/udp_test.go
@@ -38,13 +38,9 @@ func (s *connectedSuite) TestUDPEcho() {
 	mb.WriteString("This is ")
 	itm := "a russian doll containing "
 	count := 1000
-	switch runtime.GOOS {
-	case "darwin":
+	if runtime.GOOS == "darwin" {
 		// Max UDP message size is 9216 bytes
 		count = 9000 / len(itm)
-	case "windows":
-		// SO_MAX_MSG_SIZE == 8192
-		count = 8000 / len(itm)
 	}
 	for i := 0; i < count; i++ {
 		mb.WriteString(itm)

--- a/integration_test/udp_test.go
+++ b/integration_test/udp_test.go
@@ -3,22 +3,16 @@ package integration_test
 import (
 	"fmt"
 	"net"
-	"path/filepath"
 	"strings"
 	"time"
-
-	"github.com/telepresenceio/telepresence/v2/integration_test/itest"
 )
 
 func (s *connectedSuite) TestUDPEcho() {
 	ctx := s.Context()
 	require := s.Require()
 	svc := "udp-echo"
-	tag := s.Registry() + "/udp-test:" + strings.TrimPrefix(s.TelepresenceVersion(), "v")
-	testDir := filepath.Join("testdata", svc)
+	tag := "docker.io/thhal/udp-test:0.1.0"
 
-	require.NoError(itest.Run(ctx, "docker", "build", "-t", tag, testDir))
-	require.NoError(itest.Run(ctx, "docker", "push", tag))
 	require.NoError(s.Kubectl(ctx, "create", "deploy", svc, "--image", tag))
 	require.NoError(s.Kubectl(ctx, "expose", "deploy", svc, "--port", "80", "--protocol", "UDP", "--target-port", "8080"))
 	defer func() {

--- a/pkg/client/rootd/router.go
+++ b/pkg/client/rootd/router.go
@@ -76,13 +76,6 @@ func (s *session) handlePacket(c context.Context, data *buffer.Data) {
 		return
 	}
 
-	if ipHdr.PayloadLen() > buffer.DataPool.MTU-ipHdr.HeaderLen() {
-		// Packet is too large for us.
-		dlog.Error(c, "Packet exceeds MTU")
-		reply(icmp.DestinationUnreachablePacket(ipHdr, icmp.MustFragment))
-		return
-	}
-
 	if ipHdr.Version() == ipv4.Version {
 		v4Hdr := ipHdr.(ip.V4Header)
 		if v4Hdr.Flags()&ipv4.MoreFragments != 0 || v4Hdr.FragmentOffset() != 0 {

--- a/pkg/client/rootd/router.go
+++ b/pkg/client/rootd/router.go
@@ -34,22 +34,17 @@ func (s *session) routerWorker(c context.Context) error {
 		}
 	}()
 
+	buf := buffer.NewData(0x10000)
 	for atomic.LoadInt32(&s.closing) < 2 {
-		data := buffer.DataPool.Get(buffer.DataPool.MTU)
-		for {
-			n, err := s.dev.ReadPacket(data)
-			if err != nil {
-				buffer.DataPool.Put(data)
-				if c.Err() != nil || atomic.LoadInt32(&s.closing) == 2 {
-					return nil
-				}
-				return fmt.Errorf("read packet error: %w", err)
+		n, err := s.dev.ReadPacket(buf)
+		if err != nil {
+			if c.Err() != nil || atomic.LoadInt32(&s.closing) == 2 {
+				return nil
 			}
-			if n > 0 {
-				data.SetLength(n)
-				bufCh <- data
-				break
-			}
+			return fmt.Errorf("read packet error: %w", err)
+		}
+		if n > 0 {
+			bufCh <- buffer.DataPool.Copy(buf, n)
 		}
 	}
 	return nil
@@ -96,7 +91,7 @@ func (s *session) handlePacket(c context.Context, data *buffer.Data) {
 			if data == nil {
 				return
 			}
-			v4Hdr = data.Buf()
+			ipHdr = ip.V4Header(data.Buf())
 		}
 	} // TODO: similar for ipv6 using segments
 
@@ -139,9 +134,9 @@ type vifWriter struct {
 }
 
 func (w vifWriter) Write(ctx context.Context, pkt ip.Packet) (err error) {
-	dlog.Tracef(ctx, "-> TUN %s", pkt)
 	d := pkt.Data()
 	l := len(d.Buf())
+	dlog.Tracef(ctx, "-> TUN %s, len %d", pkt, l)
 	o := 0
 	for {
 		var n int

--- a/pkg/vif/buffer/pool.go
+++ b/pkg/vif/buffer/pool.go
@@ -11,12 +11,29 @@ type Pool struct {
 	MTU  int
 }
 
+// Get returns an entry from the pool if the requested size is less than or equal to
+// the pool's MTU. For sizes larger than the MTU, a new entry is allocated and returned,
+// but it is not pooled.
 func (p *Pool) Get(size int) *Data {
-	b := p.pool.Get().(*Data)
-	b.SetLength(size)
-	return b
+	if size <= p.MTU+maxIPHeader {
+		d := p.pool.Get().(*Data)
+		d.setLength(size)
+		return d
+	}
+	return NewData(size)
 }
 
+// Put will put entries with a size less than or equal to MTU back to the pool.
 func (p *Pool) Put(b *Data) {
-	p.pool.Put(b)
+	if cap(b.Buf()) == p.MTU+maxIPHeader {
+		p.pool.Put(b)
+	}
+}
+
+var DataPool = &Pool{
+	pool: sync.Pool{
+		New: func() interface{} {
+			return NewData(defaultMTU + maxIPHeader)
+		}},
+	MTU: defaultMTU,
 }

--- a/pkg/vif/buffer/pool_darwin.go
+++ b/pkg/vif/buffer/pool_darwin.go
@@ -1,9 +1,5 @@
 package buffer
 
-import (
-	"sync"
-)
-
 const PrefixLen = 4
 
 // Data on a macOS consists of two slices that share the same underlying byte array. The
@@ -22,17 +18,20 @@ func (b *Data) Buf() []byte {
 	return b.buf
 }
 
-// SetLength sets the length of this buffer. This will change the slice that Buf and Raw returns
-func (b *Data) SetLength(l int) {
-	if l > cap(b.buf) {
-		raw := b.raw
-		b.raw = make([]byte, l+PrefixLen)
-		copy(b.raw, raw)
-		b.buf = b.raw[PrefixLen:]
-	} else {
-		b.buf = b.buf[:l]
-		b.raw = b.raw[:l+PrefixLen]
-	}
+// Copy copies n bytes from the given Data buffer into a new Data which is obtained from
+// the pool, and returns the new data.
+func (p *Pool) Copy(s *Data, n int) *Data {
+	c := p.Get(n)
+	c.buf = c.buf[:n]
+	c.raw = c.raw[:n+PrefixLen]
+	copy(c.raw, s.raw)
+	return c
+}
+
+// setLength sets the length of this buffer. This will change the slice that Buf and Raw returns
+func (b *Data) setLength(l int) {
+	b.buf = b.buf[:l]
+	b.raw = b.raw[:l+PrefixLen]
 }
 
 // Raw returns this Data's raw buffer. This is the buffer that should be used by the tun.Device
@@ -42,11 +41,7 @@ func (b *Data) Raw() []byte {
 	return b.raw
 }
 
-var DataPool = &Pool{
-	pool: sync.Pool{
-		New: func() interface{} {
-			raw := make([]byte, PrefixLen+defaultMTU+maxIPHeader)
-			return &Data{buf: raw[PrefixLen:], raw: raw}
-		}},
-	MTU: defaultMTU,
+func NewData(sz int) *Data {
+	raw := make([]byte, PrefixLen+sz)
+	return &Data{buf: raw[PrefixLen:], raw: raw}
 }

--- a/pkg/vif/buffer/pool_other.go
+++ b/pkg/vif/buffer/pool_other.go
@@ -3,8 +3,6 @@
 
 package buffer
 
-import "sync"
-
 type Data struct {
 	buf []byte
 }
@@ -15,15 +13,17 @@ func (b *Data) Buf() []byte {
 	return b.buf
 }
 
-// SetLength sets the length of this buffer. This will change the slice that Buf and Raw returns
-func (b *Data) SetLength(l int) {
-	if l > cap(b.buf) {
-		buf := b.buf
-		b.buf = make([]byte, l)
-		copy(b.buf, buf)
-	} else {
-		b.buf = b.buf[:l]
-	}
+// Copy copies n bytes from the given Data buffer into a new Data which is obtained from
+// the pool, and returns the new data.
+func (p *Pool) Copy(s *Data, n int) *Data {
+	c := p.Get(n)
+	copy(c.buf, s.buf)
+	return c
+}
+
+// setLength sets the length of this buffer. This will change the slice that Buf and Raw returns
+func (b *Data) setLength(l int) {
+	b.buf = b.buf[:l]
 }
 
 // Raw returns this Data's raw buffer. This is the buffer that should be used by the tun.Device
@@ -33,10 +33,6 @@ func (b *Data) Raw() []byte {
 	return b.buf
 }
 
-var DataPool = &Pool{
-	pool: sync.Pool{
-		New: func() interface{} {
-			return &Data{buf: make([]byte, defaultMTU+maxIPHeader)}
-		}},
-	MTU: defaultMTU,
+func NewData(sz int) *Data {
+	return &Data{buf: make([]byte, sz)}
 }

--- a/pkg/vif/ip/v4header.go
+++ b/pkg/vif/ip/v4header.go
@@ -221,6 +221,8 @@ func (h V4Header) ConcatFragments(data *buffer.Data, fragsMap map[uint16][]*buff
 
 	last := V4Header(fragments[len(fragments)-1].Buf())
 	fragments = append(fragments, data)
+	fragsMap[h.ID()] = fragments
+
 	if h.FragmentOffset() < last.FragmentOffset() {
 		// Fragments didn't arrive in order. Sort them
 		sort.Slice(fragments, func(i, j int) bool {
@@ -236,7 +238,6 @@ func (h V4Header) ConcatFragments(data *buffer.Data, fragsMap map[uint16][]*buff
 	}
 
 	// Ensure that there are no holes in the fragment chain
-	lastPayload := 0
 	expectedOffset := 0
 	for _, data := range fragments {
 		eh := V4Header(data.Buf())
@@ -245,8 +246,7 @@ func (h V4Header) ConcatFragments(data *buffer.Data, fragsMap map[uint16][]*buff
 			// There's a gap. Await more fragments
 			return nil
 		}
-		lastPayload = eh.PayloadLen()
-		expectedOffset = offset + lastPayload
+		expectedOffset = offset + eh.PayloadLen()
 	}
 	totalPayload := expectedOffset
 	firstHeader := V4Header(fragments[0].Buf())
@@ -264,7 +264,7 @@ func (h V4Header) ConcatFragments(data *buffer.Data, fragsMap map[uint16][]*buff
 
 	firstHeader = fb
 	firstHeader.SetFlags(firstHeader.Flags() &^ ipv4.MoreFragments)
-	firstHeader.SetPayloadLen(firstHeader.HeaderLen() + totalPayload)
+	firstHeader.SetPayloadLen(totalPayload)
 	firstHeader.SetChecksum()
 	return final
 }

--- a/pkg/vif/udp/handler.go
+++ b/pkg/vif/udp/handler.go
@@ -33,7 +33,15 @@ func (h *timedHandler) resetIdle() bool {
 }
 
 func (h *timedHandler) Close(_ context.Context) {
-	h.remove()
+	h.idleLock.Lock()
+	if h.remove != nil {
+		h.remove()
+		h.remove = nil
+		if h.idleTimer != nil {
+			h.idleTimer.Stop()
+		}
+	}
+	h.idleLock.Unlock()
 }
 
 type handler struct {
@@ -64,7 +72,8 @@ func (h *handler) HandleDatagram(ctx context.Context, dg Datagram) {
 	case h.fromTun <- dg:
 	}
 }
-func sendUDPToTun(ctx context.Context, id tunnel.ConnID, payload []byte, toTun ip.Writer) {
+
+func createReply(id tunnel.ConnID, payload []byte) Datagram {
 	pkt := NewDatagram(HeaderLen+len(payload), id.Destination(), id.Source())
 	ipHdr := pkt.IPHeader()
 	ipHdr.SetChecksum()
@@ -75,7 +84,11 @@ func sendUDPToTun(ctx context.Context, id tunnel.ConnID, payload []byte, toTun i
 	udpHdr.SetPayloadLen(uint16(len(payload)))
 	copy(udpHdr.Payload(), payload)
 	udpHdr.SetChecksum(ipHdr)
+	return pkt
+}
 
+func sendUDPToTun(ctx context.Context, id tunnel.ConnID, payload []byte, toTun ip.Writer) {
+	pkt := createReply(id, payload)
 	defer pkt.Release()
 	if err := toTun.Write(ctx, pkt); err != nil {
 		dlog.Errorf(ctx, "!! TUN %s: %v", id, err)
@@ -84,7 +97,25 @@ func sendUDPToTun(ctx context.Context, id tunnel.ConnID, payload []byte, toTun i
 
 func (h *handler) Start(ctx context.Context) {
 	h.idleTimer = time.NewTimer(idleDuration)
+	go h.readLoop(ctx)
 	go h.writeLoop(ctx)
+}
+
+func (h *handler) readLoop(ctx context.Context) {
+	defer h.Close(ctx)
+	for ctx.Err() == nil {
+		m, err := h.stream.Receive(ctx)
+		if err != nil {
+			return
+		}
+		switch m.Code() {
+		case tunnel.DialOK:
+		case tunnel.DialReject, tunnel.Disconnect:
+			return
+		case tunnel.Normal:
+			sendUDPToTun(ctx, h.id, m.Payload(), h.toTun)
+		}
+	}
 }
 
 func (h *handler) writeLoop(ctx context.Context) {
@@ -100,8 +131,8 @@ func (h *handler) writeLoop(ctx context.Context) {
 				dg.Release()
 				return
 			}
-			dlog.Debugf(ctx, "<- TUN %s", dg)
-			dlog.Debugf(ctx, "-> MGR %s", dg)
+			dlog.Tracef(ctx, "<- TUN %s", dg)
+			dlog.Tracef(ctx, "-> MGR %s", dg)
 			udpHdr := dg.Header()
 			err := h.stream.Send(ctx, tunnel.NewMessage(tunnel.Normal, udpHdr.Payload()))
 			dg.Release()

